riscv/sim.h

```diff
@@ -70,6 +70,7 @@ class sim_t : public htif_t, public simif_t
   std::string dts;
   std::string dtb;
   bool dtb_enabled;
+  std::vector<std::shared_ptr<abstract_device_t>> devices;
```

Comment:
Since the bus already needs to maintain a list of all devices, can we avoid some redundancy by just calling `bus.tick()` and having that tick everything?

Reply:
I think we can make the bus tick everything, but the list of devices should still be maintained by `sim_t`, to avoid dealing with shared ownership, and to avoid introducing `shared_ptrs` to more places.

Reply:
The ticking was less my concern; it was more that we are appending every device to both the bus and the device list. If there’s some other way to be DRY about that, I’d be OK with it.

Reply:
If you are ok with changing device pointers to  `shared_ptr` everywhere, then we can DRY it. 

The problem is that some devices can be solely owned by the bus (rom, plugins), while others need to be accessed through the sim interface as well.

Reply:
Nvm I see what you are saying

Reply:
Actually I think the current set up is best.
`sim_t.devices` maintains ownership of the devices under sim_t, while `sim_t.bus.devices` tracks which devices are managers under the bus.

The code is already pretty DRY I think, only the boot_rom has some special handling since it is instantiated at `sim_t::reset`, not the constructor.

Reply:
Yeah, I'm OK with only the boot ROM being a special case in this respect, since, as you point out, it is a special case.